### ISSUE TITLE
Ajax update of empty EL in script fails

### ISSUE THREAD

**Issue by jthalmai** [eclipse-ee4j/mojarra#4647]
## Environment Details

* WildFly 18.0.0
* JDK version:8
* OS:windows
* Database:none

## Problem Description

In the example below there is a <script> tag that contains some Expression Language usages. If the EL evaluates to some String, this is fine. But if it evaluates to an empty String or null and the part of the page that contains the <script> is updated with Ajax, this fails in WildFly 18 (Mojarra 2.3.9.SP03). It worked in WildFly 17.0.1 (Mojarra 2.3.9.SP02).

In the example, you get the error, once you press the button.

xhtml:

    <h:form id="form">

    <h:panelGroup id="js">
    <script type="text/javascript">
    console.log("not empty String: #{testView.notEmptyString}"); // this works
    console.log("empty String: #{testView.emptyString}"); // this fails
    console.log("null: #{testView.nullString}"); // this fails
    </script>
    </h:panelGroup>

    <h:commandButton value="update">
    <f:ajax render="form:js" />
    </h:commandButton>

    </h:form>

TestView bean:

    public String getEmptyString() {
        return "";
    }
	
    public String getNotEmptyString() {
        return "some string";
    }
	
    public String getNullString() {
        return null;
    }

Stacktrace:

```
Error Rendering View[/elscript.xhtml]: java.lang.ArrayIndexOutOfBoundsException: -2
	at com.sun.faces.renderkit.html_basic.HtmlResponseWriter.writeUnescapedCData(HtmlResponseWriter.java:1474)
	at com.sun.faces.renderkit.html_basic.HtmlResponseWriter.writeText(HtmlResponseWriter.java:961)
	at javax.faces.context.ResponseWriterWrapper.writeText(ResponseWriterWrapper.java:296)
	at javax.faces.context.ResponseWriterWrapper.writeText(ResponseWriterWrapper.java:296)
	at javax.faces.context.ResponseWriterWrapper.writeText(ResponseWriterWrapper.java:296)
	at com.sun.faces.facelets.el.ELText$ELTextVariable.writeText(ELText.java:218)
	at com.sun.faces.facelets.el.ELText$ELTextComposite.writeText(ELText.java:123)
	at com.sun.faces.facelets.compiler.TextInstruction.write(TextInstruction.java:44)
	at com.sun.faces.facelets.compiler.UIInstructions.encodeBegin(UIInstructions.java:41)
	at com.sun.faces.renderkit.html_basic.HtmlBasicRenderer.encodeRecursive(HtmlBasicRenderer.java:276)
	at com.sun.faces.renderkit.html_basic.GroupRenderer.encodeChildren(GroupRenderer.java:90)
	at javax.faces.component.UIComponentBase.encodeChildren(UIComponentBase.java:566)
	at javax.faces.component.UIComponent.encodeAll(UIComponent.java:1647)
	at com.sun.faces.context.PartialViewContextImpl$PhaseAwareVisitCallback.visit(PartialViewContextImpl.java:615)
	at com.sun.faces.component.visit.PartialVisitContext.invokeVisitCallback(PartialVisitContext.java:159)
	at javax.faces.component.UIComponent.visitTree(UIComponent.java:1456)
	at javax.faces.component.UIComponent.visitTree(UIComponent.java:1468)
	at javax.faces.component.UIForm.visitTree(UIForm.java:355)
	at javax.faces.component.UIComponent.visitTree(UIComponent.java:1468)
	at javax.faces.component.UIComponent.visitTree(UIComponent.java:1468)
	at com.sun.faces.context.PartialViewContextImpl.processComponents(PartialViewContextImpl.java:400)
	at com.sun.faces.context.PartialViewContextImpl.processPartial(PartialViewContextImpl.java:319)
	at org.primefaces.context.PrimePartialViewContext.processPartial(PrimePartialViewContext.java:65)
	at javax.faces.component.UIViewRoot.encodeChildren(UIViewRoot.java:1101)
	at javax.faces.component.UIComponent.encodeAll(UIComponent.java:1647)
	at com.sun.faces.application.view.FaceletViewHandlingStrategy.renderView(FaceletViewHandlingStrategy.java:442)
	at com.sun.faces.application.view.MultiViewHandler.renderView(MultiViewHandler.java:170)
	at javax.faces.application.ViewHandlerWrapper.renderView(ViewHandlerWrapper.java:132)
	at javax.faces.application.ViewHandlerWrapper.renderView(ViewHandlerWrapper.java:132)
	at com.sun.faces.lifecycle.RenderResponsePhase.execute(RenderResponsePhase.java:102)
	at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:76)
	at com.sun.faces.lifecycle.LifecycleImpl.render(LifecycleImpl.java:199)
	at javax.faces.webapp.FacesServlet.executeLifecyle(FacesServlet.java:708)
	at javax.faces.webapp.FacesServlet.service(FacesServlet.java:451)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at io.opentracing.contrib.jaxrs2.server.SpanFinishingFilter.doFilter(SpanFinishingFilter.java:52)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:68)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.deployment.GlobalRequestControllerHandler.handleRequest(GlobalRequestControllerHandler.java:68)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:269)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:78)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:133)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:130)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction.lambda$create$0(SecurityContextThreadSetupAction.java:105)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1504)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1504)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1504)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1504)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:249)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:78)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:99)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:376)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1982)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
	at java.lang.Thread.run(Unknown Source)
```

**Comment by arjantijms** [comment]
Somewhat of a standard reply, but did you test with the latest 2.3.13 release and/or the master version?

**Comment by thomaswoeckinger** [comment]
Same here broken at 2.3.13 and 2.3.14, still working with 2.3.12

**Comment by cmunier** [comment]
From my perspective, this is a **severe problem** that should be handled with **high priority**.

We are using a component framework where it is common that composite component attributes are used as parameters for encapsulated JavaScript for added functionality. It is a common case that the default value of an optional attribute is empty. So there you now always end up with an Exception. To my mind there is no useful workaround.

**Comment by thomaswoeckinger** [comment]
Totally confirm with cmunier, the current state does not allow an update from 2.3.12 to 2.3.14, from my opinion this is a blocking bug!

**Comment by entity-development** [comment]
Seems to be fixed in https://repository.jboss.org/nexus/content/repositories/public/com/sun/faces/jsf-impl/2.3.9.SP06/

**Comment by mattwltr** [comment]
Seems to be fixed in https://repository.jboss.org/nexus/content/repositories/public/com/sun/faces/jsf-impl/2.3.9.SP06/

**Comment by Jarannn** [comment]
> 
> 
> Seems to be fixed in https://repository.jboss.org/nexus/content/repositories/public/com/sun/faces/jsf-impl/2.3.9.SP06/

I can confirm what mattwltr reported: just copying the patched HtmlResponseWriter (I think it was fixed by https://github.com/jboss/mojarra/pull/57) did resolve my problems with the empty EL evaluation issue. (Payara 5.201)

**Comment by scolapasta** [comment]
@Jarannn would you be willing to explain in a little more detail what you copied from where to where? We are also running on payara 5.201 (with jsf2.3.14) and will likely need to patch to get around this.

**Comment by scolapasta** [comment]
@jaram thanks that worked for us

So, will this fix get into an upcoming official version? We are working un upgrading our app and moving to a Payara5 in production, but right now, if we do, we'll have to use a patched version of the jar (and make that part of the installation of our software).

**Comment by BalusC** [comment]
Confirmed fixed in 2.3.16 (and 2.3.17), I only cannot find associated integration test.

**Comment by BalusC** [comment]
Found it: #4925

Closing off.
